File: 백준/Silver/11558. The Game of Death/The Game of Death.c
#include <stdio.h>
#include <stdlib.h>

/*
그냥 어렵게 생각할 필요없고 반복문으로 풀어도 된다.
구현문제라고 생각하고 그냥 지명한 숫자대로 따라가면 된다.
*/

int main(void)
{
    int T;
    int n; // 주경이는 n번이다.
    int * call;
    int i, j;

    int callCnt; // 지명 횟수.
    int callIdx; // 지명권한을 가지고 있는 사람번호 - 1
    int findFlag;

    
    scanf("%d", &T);

    for(i = 0; i < T; i++)
    {
        scanf("%d", &n);

        call = (int*)malloc(sizeof(int) * n);
        for(j = 0; j < n; j++)
            scanf("%d", &call[j]);

        callCnt = 1; // 무조건 한 번은 지명하므로 1부터 시작.
        callIdx = 0;
        findFlag = 0;

 
        while(callCnt <= n)
        {
            if(call[callIdx] != n)
            {
                callCnt++;
                callIdx = call[callIdx] - 1; // 지명 권한(?)이 다음 사람에게 넘어감.
            }
            else
            {
                findFlag = 1;
                break;
            }
        }

        printf("%d\n", findFlag==1 ? callCnt : 0);
        
        free(call);
    }
    

    return 0;
}

File: 백준/Silver/11558. The Game of Death/The Game of Death.py
T = int(input())

for test_case in range(1, T+1):
    n = int(input()) # 플레이어 수. 주경이는 n번.
    call = []
    k = 1 # 지목 횟수
    findFlag = 0
    idx = 0 # call 리스트의 인덱스. 현재 사람의 번호 저장.

    # 입력받기
    for _ in range(n):
        call.append(int(input()))
    
    # 사람이 n명 있으므로 n번의 지목 횟수를 채우면
    # 더이상 탐색을 할 필요가 없어진다.
    while k <= n:
        # 현재 사람이 지목한 사람이 주경이가 아니라면
        if call[idx] != n:
            k += 1 # 지목횟수 증가
            idx = call[idx] - 1 # 다음 사람으로 넘어감
        else:
            findFlag = 1 # 주경이가 걸렸음을 의미
            break

    if findFlag == 0:
        print("0")
    else:
        print(k)